File: frontend/src/utils/tradingview.ts
export function getTradingViewUrl(assetSymbol: string, customUrl?: string): string {
  // If custom URL is provided, use it
  if (customUrl) {
    return customUrl;
  }

  // Otherwise, generate default URL
  // Parse asset_symbol to get code and country_code
  const [code, countryCode = 'xpar'] = assetSymbol.split(':');

  // Map country codes to TradingView exchanges
  const exchangeMap: Record<string, string> = {
    'xpar': 'EURONEXT',
    'xnas': 'NASDAQ',
    'xnys': 'NYSE',
    'xlon': 'LSE',
    'xfra': 'FWB',
    'xetr': 'XETR',
  };

  const exchange = exchangeMap[countryCode.toLowerCase()] || 'EURONEXT';
  const symbol = code.toUpperCase();

  return `https://www.tradingview.com/chart/?symbol=${exchange}:${symbol}`;
}

File: utils/tradingview.py
from typing import Optional

EXCHANGE_MAP = {
    "xpar": "EURONEXT",
    "xnas": "NASDAQ",
    "xnys": "NYSE",
    "xlon": "LSE",
    "xfra": "FWB",
    "xetr": "XETR",
}

_DEFAULT_EXCHANGE = "EURONEXT"


def build_tradingview_url(code: str, country_code: Optional[str]) -> str:
    exchange = EXCHANGE_MAP.get(
        (country_code or "").lower(), _DEFAULT_EXCHANGE
    )
    return (
        f"https://www.tradingview.com/chart/?symbol={exchange}:{code.upper()}"
    )


def build_tradingview_url_from_symbol(symbol: str) -> str:
    if ":" in symbol:
        code, country_code = symbol.split(":", 1)
    else:
        code, country_code = symbol, None
    return build_tradingview_url(code, country_code)
